text/0088-metric-instrument-optional-refinements.md

```diff
@@ -0,0 +1,306 @@
+# Metric Instruments
+
+Removes the optional semantic declarations `Monotonic` and `Absolute`
+for metric instruments, declares the Measure and Observer instruments
+as _foundational_, and introduces a process for standardizing new
+instrument _refinements_.
+
+## Motivation
+
+With the removal of Gauge instruments and the addition of Observer
+instruments in the specification, the existing `Monotonic` and
+`Absolute` options began to create confusion.  For example, a Counter
+instrument is used for capturing changes in a Sum, and we could say
+that non-negative-valued metric events define a monotonic Counter, in
+the sense that its Sum is monotonic.  The confusion arises, in this
+case, because `Absolute` refers to the captured values, whereas
+`Monotonic` refers to the semantic output.
+
+From a different perspective, Counter instruments might be treated as
+refinements of the Measure instrument.  Whereas the Measure instrument
+is used for capturing all-purpose synchronous measurements, the
+Counter instrument is used specifically for synchronously capturing
+measurements of changes in a sum, therefore it uses `Add()` instead of
+`Record()`, and it specifies `Sum` as the standard aggregation.
+
+What this illustrates is that we have modeled this space poorly.  This
+does not propose to change any existing metrics APIs, only our
+understanding of the three instruments currently included in the
+specification: Measure, Observer, and Counter.
+
+## Explanation
+
+The Measure and Observer instrument are defined as _foundational_
+here, in the sense that any kind of metric instrument must reduce to
+one of these.  The foundational instruments are unrestricted, in the
+sense that metric events support any numerical value, positive or
+negative, zero or infinity.
+
+The distinction between the two foundational instruments is whether
+they are synchronous.  Measure instruments are called synchronously by
+the user, while Observer instruments are called asynchronously by the
+implementation.  Synchronous instruments (Measure and its refinements)
+have three calling patterns (_Bound_, _Unbound_, and _Batch_) to
+capture measurements.  Asynchronous instruments (Observer and its
+refinements) use callbacks to capture measurements.
+
+All measurement APIs produce metric events consisting of [timestamp,
+instrument descriptor, label set, and numerical
+value](api-metrics.md#metric-event-format).  Synchronous instrument
+events additionally have [Context](api-context.md), describing
+properties of the associated trace and distributed correlation values.
+
+### Last-value relationship
+
+Observer instruments have a well-defined _last value_ measured by the
+instrument that can be useful in defining aggregations.  To maintain
+this property, we impose this requirement: two or more calls to
+`Observe()` in a single Observer callback invocation are treated as
+duplicates of each other, and the last call to `Observe()` wins.
```

Comment:
Are the calls to `Observe()` considered duplicate even if the calls are with a different labelset for each?

Reply:
I have this question too.  In the current spec, there is a sentence which says "_A single Observer instrument callback can capture multiple metric events associated with different label sets._" which seems to imply they are not duplicates if the label sets are different.

However, for a metric which has [WithRecommendedKeys](https://github.com/open-telemetry/opentelemetry-specification/blob/4ab7c2fb4964c14fe972b03b5bb6825573bafb24/specification/api-metrics.md#withrecommendedkeys-declaration-on-metric-instruments), are two different observations with nominally different labelsets at the point of observation treated as duplicates if the `WithRecommendedKeys` option collapses those labelsets into the same group?

Reply:
There is a motion to remove `WithRecommendedKeys` and the associated "defaultkeys" Batcher (a.k.a. Integrator).  Still, the kind of aggregation that happens when combining two distinct LabelSets into a group is referred to as spatial aggregation here, and it would apply whatever the ordinary aggregation is for the instrument.

This is why "last value" is referred to as a relationship given by Observers, not an aggregation. You can configure an asynchronous instrument to aggregate the last value using by adding values (as for `DeltaObserver` in this proposal), or by combining the distribution (as for the unrefined `Observer` in this proposal). See also this remark: https://github.com/open-telemetry/oteps/pull/88#discussion_r398204203

Reply:
See https://github.com/open-telemetry/opentelemetry-specification/issues/463.